apache/superset
# feat(rbac): add customizable related filters

### SUMMARY
This is a continuation of #21515, which made it possible to exclude specific users from dropdown lists, to make it possible to add arbitrary filter conditions to both the user and role dropdowns. The primary reason for this is to be able to restrict the available roles in the Dashboard role dropdown when using `DASHBOARD_RBAC`, as users may unintentionally give too wide access to dashboards by sharing with a too generic role. Similarly, there may be a need to restrict to whom a user should be able to give ownership of an object.

To demonstrate, by adding the following to `superset_config.py`, the owners dropdown now only contains the entry `admin`, and the roles dropdown only contains entries starting with the letter "A":

```python
def role_filter(query):
    from superset import security_manager

    role_model = security_manager.role_model
    filters = [role_model.name.like("A%")]
    return query.filter(or_(*filters))


def user_filter(query):
    from superset import security_manager

    user_model = security_manager.user_model
    filters = [user_model.username == "admin"]
    return query.filter(or_(*filters))


EXTRA_RELATED_QUERY_FILTERS = {
    "role": role_filter,
    "user": user_filter,
}
```

Now users only displays "Admin":

<img width="887" alt="image" src="https://user-images.githubusercontent.com/33317356/209462025-fb9c46be-b772-461d-bf8b-473aeeda0760.png">

Similarly roles only contains "Admin" and "Alpha":

<img width="884" alt="image" src="https://user-images.githubusercontent.com/33317356/209462080-af7f9786-fb74-41e9-8906-052fc8e148e9.png">

### TESTING INSTRUCTIONS
1. Add the above to `superset_config.py`
2. Go to dashboard, chart and dataset CRUD views and ensure that the owners and roles dropdowns only contains the expected values.

### ADDITIONAL INFORMATION
